pyscf/pbc/dft/multigrid/multigrid_pair.py

```diff
@@ -0,0 +1,1358 @@
+        nelec = nelec[0]
+        excsum = excsum[0]
+    log.debug('Multigrid exc %s  nelec %s', excsum, nelec)
+
+    kpts_band, input_band = _format_kpts_band(kpts_band, kpts), kpts_band
+    if xctype == 'LDA':
+        if with_j:
+            for s in range(2):
+                wv_freq[:,s,0] += vG.reshape(nset,*mesh)
+        veff = _get_j_pass2(mydf, wv_freq, kpts_band, verbose=log)
+    elif xctype == 'GGA':
+        if with_j:
+            #wv_freq[:,:,0] += vG.reshape(nset,*mesh)
+            for s in range(2):
+                wv_freq[:,s,0] = lib.add(wv_freq[:,s,0], vG.reshape(nset,*mesh), out=wv_freq[:,s,0])
+        if GGA_METHOD.upper() == 'FFT':
+            veff = _get_j_pass2(mydf, wv_freq, kpts_band, verbose=log)
+        else:
+            veff = _get_gga_pass2(mydf, wv_freq, kpts_band, hermi=hermi, verbose=log)
+    wv_freq = None
+    veff = _format_jks(veff, dm_kpts, input_band, kpts)
+
+    if return_j:
+        vj = _get_j_pass2(mydf, vG, kpts_band, verbose=log)
+        vj = _format_jks(veff, dm_kpts, input_band, kpts)
+    else:
+        vj = None
+    vG = None
+
+    veff = lib.tag_array(veff, ecoul=ecoul, exc=excsum, vj=vj, vk=None)
+    return nelec, excsum, veff
+
+def get_veff_ip1(mydf, dm_kpts, xc_code=None, kpts=np.zeros((1,3)), kpts_band=None, spin=0):
+    cell = mydf.cell
+    dm_kpts = lib.asarray(dm_kpts, order='C')
+    dms = _format_dms(dm_kpts, kpts)
+    nset, nkpts, nao = dms.shape[:3]
+    kpts_band = _format_kpts_band(kpts_band, kpts)
+    if spin == 1:
+        nset //= 2
+
+    mesh = mydf.mesh
+    ngrids = np.prod(mesh)
+    ni = mydf._numint
+    xctype = ni._xc_type(xc_code)
+    if xctype == 'LDA':
+        deriv = 0
+    elif xctype == 'GGA':
+        deriv = 1
+    rhoG = _eval_rhoG(mydf, dm_kpts, hermi=1, kpts=kpts_band, deriv=deriv)
+    if spin == 1:
+        rhoG = rhoG.reshape(nset,2,-1,ngrids)
+    # cache rhoG for core density gradients
+    mydf.rhoG = rhoG
+
+    coulG = tools.get_coulG(cell, mesh=mesh)
+    vG = np.empty((nset,ngrids), dtype=np.result_type(rhoG, coulG))
+    for i in range(nset):
+        if spin == 0:
+            vG[i] = lib.multiply(rhoG[i,0], coulG, out=vG[i])
+        elif spin == 1:
+            tmp = lib.add(rhoG[i,0,0], rhoG[i,1,0])
+            vG[i] = lib.multiply(tmp, coulG, out=vG[i])
+
+    if mydf.vpplocG_part1 is not None and not mydf.pp_with_erf:
+        for i in range(nset):
+            vG[i] = lib.add(vG[i], mydf.vpplocG_part1, out=vG[i])
+
+    weight = cell.vol / ngrids
+
+    # *(1./weight) because rhoR is scaled by weight in _eval_rhoG.  When
+    # computing rhoR with IFFT, the weight factor is not needed.
+    rhoR = tools.ifft(rhoG.reshape(-1,ngrids), mesh).real * (1./weight)
+    if spin == 0:
+        rhoR = rhoR.reshape(nset,-1,ngrids)
+    elif spin == 1:
+        rhoR = rhoR.reshape(nset,2,-1,ngrids)
+
+    wv_freq = []
+    for i in range(nset):
+        exc, vxc = ni.eval_xc(xc_code, rhoR[i], spin=spin, deriv=1)[:2]
```

Comment:
It's recommended to use the eval_xc_eff method like https://github.com/pyscf/pyscf/blob/881dbbebcbc084ba2ba8fe968ff03fd34d7dfdf8/pyscf/dft/numint.py#L1138 . the vxc for LDA, GGA, MGGA can be handled with almost the same code. Then the `_rks_gga_wv{0,1,2}` functions are not needed.

Reply:
I may update this in a future PR as the G-space approximation to the GGA potential may not be compatible.